🐛 BUG: fetching bound registered workers in local mode fails with timeout

### What version of `Wrangler` are you using?

2.1.6

### What operating system are you using?

M1 Pro macOS 12.6

### Describe the Bug

When running `wrangler pages dev`, after some delay (presumably the timeout mentioned in the error) this error is being output several times per second until the server is stopped.

```
▲ [WARNING] Failed to get worker definitions TypeError: fetch failed

      at Object.processResponse
  (/Users/andre/code/wrangler-bug/node_modules/wrangler/wrangler-dist/cli.js:8846:27)
      at /Users/andre/code/wrangler-bug/node_modules/wrangler/wrangler-dist/cli.js:9178:42
      at node:internal/process/task_queues:140:7
      at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
      at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    cause: ConnectTimeoutError: Connect Timeout Error
        at onConnectTimeout
  (/Users/andre/code/wrangler-bug/node_modules/wrangler/wrangler-dist/cli.js:2870:29)
        at /Users/andre/code/wrangler-bug/node_modules/wrangler/wrangler-dist/cli.js:2828:51
        at Immediate._onImmediate
  (/Users/andre/code/wrangler-bug/node_modules/wrangler/wrangler-dist/cli.js:2859:13)
        at process.processImmediate (node:internal/timers:471:21) {
      code: 'UND_ERR_CONNECT_TIMEOUT'
    }
  }
```

### steps to reproduce:

 - `npx create-remix@latest` 
     - name
     - Just The Basics
     - Cloudflare Pages
     - Typescript
     - Y
 - `npm run dev` 
 - wait 10-15 seconds.

Note: I do not have wrangler installed globally, and have tried running this same command after running `npx wrangler login` and the error is still output.

I'm not sure it is the same bug, but I'm getting this:

```
Failed to get worker definitions TypeError: fetch failed

      at fetch10
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:13582:17)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async getRegisteredWorkers
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:120431:22)
      at async getBoundRegisteredWorkers
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:120449:29)
  {
    cause: ConnectTimeoutError: Connect Timeout Error
        at onConnectTimeout
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:6526:29)
        at
  /Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:6484:51
        at Immediate._onImmediate
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:6515:13)
        at process.processImmediate (node:internal/timers:471:21) {
      code: 'UND_ERR_CONNECT_TIMEOUT'
    }
  }
```
Then I can kill the worker, either via kill -9 <pid> or control+C or pressing `x`, and even then I get this log out:
```
[ERROR] Failed to unregister worker TypeError: fetch failed

      at fetch10
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:13582:17)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async unregisterWorker
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:120418:5)
      at async Promise.allSettled (index 0) {
    cause: ConnectTimeoutError: Connect Timeout Error
        at onConnectTimeout
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:6526:29)
        at
  /Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:6484:51
        at Immediate._onImmediate
  (/Users/blah/backend/cloudflare-workers/account/node_modules/wrangler/wrangler-dist/cli.js:6515:13)
        at process.processImmediate (node:internal/timers:471:21) {
      code: 'UND_ERR_CONNECT_TIMEOUT'
    }
  }
```
It takes a little bit to stop the server. And I'm starting the worker using `wrangler dev --local --persist` and other workers without --persist (as they don't need the d1 db).

--- a/.changeset/gentle-yaks-repeat.md
+++ b/.changeset/gentle-yaks-repeat.md
@@ -1,0 +1,5 @@
+---
+"wrangler": major
+---
+
+Fix: fix local registry server closed

--- a/.changeset/gentle-yaks-repeat.md
+++ b/.changeset/gentle-yaks-repeat.md
@@ -3,3 +3,5 @@
 ---
 
 Fix: fix local registry server closed
+
+Closes #1920 

--- a/.changeset/gentle-yaks-repeat.md
+++ b/.changeset/gentle-yaks-repeat.md
@@ -4,4 +4,6 @@
 
 Fix: fix local registry server closed
 
-Closes [#1920](https://github.com/cloudflare/workers-sdk/issues/1920) 
+Closes [#1920](https://github.com/cloudflare/workers-sdk/issues/1920). Sometimes start the local dev server will kill
+the devRegistry server so that the devRegistry server can't be used. We can listen the devRegistry server close event
+and reset server to `null`. When registerWorker is called, we can check if the server is `null` and start a new server.

--- a/.changeset/gentle-yaks-repeat.md
+++ b/.changeset/gentle-yaks-repeat.md
@@ -4,4 +4,4 @@
 
 Fix: fix local registry server closed
 
-Closes #1920 
+Closes [#1920](https://github.com/cloudflare/workers-sdk/issues/1920) 

--- a/.changeset/gentle-yaks-repeat.md
+++ b/.changeset/gentle-yaks-repeat.md
@@ -1,5 +1,5 @@
 ---
-"wrangler": major
+"wrangler": patch
 ---
 
 Fix: fix local registry server closed

--- a/packages/wrangler/src/__tests__/unstableDev.test.ts
+++ b/packages/wrangler/src/__tests__/unstableDev.test.ts
@@ -1,5 +1,9 @@
 import { fetch } from "undici";
-import { stopWorkerRegistry, registerWorker, startWorkerRegistry } from "../dev-registry";
+import {
+	stopWorkerRegistry,
+	registerWorker,
+	startWorkerRegistry,
+} from "../dev-registry";
 
 jest.unmock("undici");
 
@@ -8,17 +12,17 @@ jest.unmock("undici");
  */
 describe("unstable devRegistry testing", () => {
 	afterAll(async () => {
-		await stopWorkerRegistry()
+		await stopWorkerRegistry();
 	});
 
 	it("should start the devRegistry if the devRegistry not start", async () => {
 		await registerWorker("test", {
 			port: 6789,
-			protocol: 'http',
-			host: 'localhost',
-			mode: 'local',
-			durableObjects: [{ name: 'testing', className: 'testing' }]
-		})
+			protocol: "http",
+			host: "localhost",
+			mode: "local",
+			durableObjects: [{ name: "testing", className: "testing" }],
+		});
 		const resp = await fetch("http://localhost:6284/workers");
 		if (resp) {
 			const parsedResp = (await resp.json()) as {
@@ -28,21 +32,21 @@ describe("unstable devRegistry testing", () => {
 		}
 	});
 
-	it('should not restart the devRegistry if the devRegistry already start', async () => {
-		await startWorkerRegistry()
+	it("should not restart the devRegistry if the devRegistry already start", async () => {
+		await startWorkerRegistry();
 
 		await fetch("http://localhost:6284/workers/init", {
-			method: 'POST',
-			body: JSON.stringify({})
+			method: "POST",
+			body: JSON.stringify({}),
 		});
 
 		await registerWorker("test", {
 			port: 6789,
-			protocol: 'http',
-			host: 'localhost',
-			mode: 'local',
-			durableObjects: [{ name: 'testing', className: 'testing' }]
-		})
+			protocol: "http",
+			host: "localhost",
+			mode: "local",
+			durableObjects: [{ name: "testing", className: "testing" }],
+		});
 
 		const resp = await fetch("http://localhost:6284/workers");
 		if (resp) {
@@ -53,5 +57,5 @@ describe("unstable devRegistry testing", () => {
 			expect(parsedResp.init).toBeTruthy();
 			expect(parsedResp.test).toBeTruthy();
 		}
-	})
+	});
 });

--- a/packages/wrangler/src/__tests__/unstableDev.test.ts
+++ b/packages/wrangler/src/__tests__/unstableDev.test.ts
@@ -1,0 +1,57 @@
+import { fetch } from "undici";
+import { stopWorkerRegistry, registerWorker, startWorkerRegistry } from "../dev-registry";
+
+jest.unmock("undici");
+
+/**
+ * Sometimes the devRegistry killed by some reason, the register worker will to restart it.
+ */
+describe("unstable devRegistry testing", () => {
+	afterAll(async () => {
+		await stopWorkerRegistry()
+	});
+
+	it("should start the devRegistry if the devRegistry not start", async () => {
+		await registerWorker("test", {
+			port: 6789,
+			protocol: 'http',
+			host: 'localhost',
+			mode: 'local',
+			durableObjects: [{ name: 'testing', className: 'testing' }]
+		})
+		const resp = await fetch("http://localhost:6284/workers");
+		if (resp) {
+			const parsedResp = (await resp.json()) as {
+				test: unknown;
+			};
+			expect(parsedResp.test).toBeTruthy();
+		}
+	});
+
+	it('should not restart the devRegistry if the devRegistry already start', async () => {
+		await startWorkerRegistry()
+
+		await fetch("http://localhost:6284/workers/init", {
+			method: 'POST',
+			body: JSON.stringify({})
+		});
+
+		await registerWorker("test", {
+			port: 6789,
+			protocol: 'http',
+			host: 'localhost',
+			mode: 'local',
+			durableObjects: [{ name: 'testing', className: 'testing' }]
+		})
+
+		const resp = await fetch("http://localhost:6284/workers");
+		if (resp) {
+			const parsedResp = (await resp.json()) as {
+				test: unknown;
+				init: unknown;
+			};
+			expect(parsedResp.init).toBeTruthy();
+			expect(parsedResp.test).toBeTruthy();
+		}
+	})
+});

--- a/packages/wrangler/src/dev-registry.ts
+++ b/packages/wrangler/src/dev-registry.ts
@@ -98,7 +98,7 @@ export async function startWorkerRegistry() {
 		/**
 		 * The registry server may close. Reset the server to null for restart.
 		 */
-		server.on('close', () => {
+		server.on("close", () => {
 			server = null;
 		});
 	}

--- a/packages/wrangler/src/dev-registry.ts
+++ b/packages/wrangler/src/dev-registry.ts
@@ -94,6 +94,13 @@ export async function startWorkerRegistry() {
 				throw err;
 			}
 		});
+
+		/**
+		 * The registry server may close. Reset the server to null for restart.
+		 */
+		server.on('close', () => {
+			server = null;
+		});
 	}
 }
 
@@ -112,6 +119,10 @@ export async function registerWorker(
 	name: string,
 	definition: WorkerDefinition
 ) {
+	/**
+	 * Prevent the dev registry be closed.
+	 */
+	await startWorkerRegistry();
 	try {
 		return await fetch(`${DEV_REGISTRY_HOST}/workers/${name}`, {
 			method: "POST",

